### ISSUE TITLE
Producer Kafka errors and exceptions

### ISSUE THREAD

**Issue by vin18** [confluentinc/confluent-kafka-python#311]
I am trying to find conditions in which KafkaError / KafkaException is thrown. 
Are errors not throw and just set to the error object in callbacks? e.g. ALL_BROKERS_DOWN error is just set and not thrown? 
```
def get_producer_client(self):
        return Producer({
            'client.id': 'leonardo',
            'bootstrap.servers': get_config('kafka', 'bootstrap.servers'),
            # Use case: only one connection and only one message will be produced every few hours
            'batch.num.messages': 1,
            'queue.buffering.max.ms': 50,
            'retries': 5,
            'default.topic.config': {
                'acks': 'all',
                'message.timeout.ms': self.local_msg_timeout_ms},
            'retry.backoff.ms': 1000,
            'max.in.flight.requests.per.connection': 1,
            'produce.offset.report': False,
            'error_cb': _error_cb})

```
```
    def sync_produce(self, message):
        try:
            self.client.produce(self.topic,
                                bytes(json.dumps(message), 'utf-8'),
                                callback=self.delivery_callback)
            log('info', 'Message Produced on Kafka')
            # Block till the message ack is received
            log('info', 'Blocked on Kafka message ack')
            self.client.poll(0)
            self.client.flush(self.flush_timeout_ms)
        except KafkaException as e:
            print("ERROR")
```

**Comment by edenhill** [comment]
Exceptions are typically only thrown on invalid use of the APIs,

For recoverable runtime errors, such as ALL_BROKERS_DOWN, it would be counter-productive to raise an exception since the problem is most likely temporary and the client will recover automatically.

General errors are thus emitted to the error_cb, while producer delivery errors are emitted on the produce callback.

**Comment by vin18** [comment]
Continuing from here: 
https://github.com/edenhill/librdkafka/issues/1685

The callback for produce is not being called. Is there a way to explicitly check how many messages are there on the local_queue ?

**Comment by vin18** [comment]
producer.flush return the size of the internal message queue right?
I can use that to check the count of pending messages ?
https://github.com/confluentinc/confluent-kafka-python/blob/41e9d546a44d4d43f00f80a3b7b8fa583cdb7cbb/confluent_kafka/src/Producer.c#L440

**Comment by edenhill** [comment]
Yes, do note that it also accounts for outstanding ops (events) enqueued for the application to poll.
So this value should not be treated as an exact measure of the number of outstanding messages, but rather a level thing which is at least the number of outstanding messages, might be more.

**Comment by aswinjoseroy** [comment]
@edenhill what is the best practice to deal with callback errors? For eg. I have a consumer which pushes messages to another topic after processing. If the produced messages fail after consumer commits (I can raise an error within the callback method and log it), what is a clean approach to reprocess the lost messages? 

Let me know if I should raise a separate issue for this.

**Comment by edenhill** [comment]
This is what Kafka Streams sets out to solve, using cross-producer-consumer transactions.

Since transaction support is not available in librdkafka, you will need to make something with less strong guarantees, like committing the offset after the secondary produce is successful.
This gives at-least-once guarantees.
